src/internal/coredb/projects.go

```diff
@@ -0,0 +1,96 @@
+package coredb
+
+import (
+	"context"
+	"fmt"
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	"time"
+)
+
+const (
+	insertProjectQuery     = "INSERT INTO core.projects (name, description, created_at, updated_at) VALUES ($1,$2,$3,$4);"
+	updateProjectBaseQuery = "UPDATE core.projects SET name = $1, description = $2, created_at = $3, updated_at = $4"
+)
+
+// CreateProject creates an entry in the core.projects table.
+func CreateProject(ctx context.Context, tx *pachsql.Tx, project *pfs.ProjectInfo) error {
+	//todo: this can be swapped out with postgres' created_at trigger later.
+	createdAt, err := types.TimestampFromProto(project.CreatedAt)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	_, err = tx.ExecContext(ctx, insertProjectQuery, project.Project.Name, project.Description, createdAt, nil)
+	//todo: insert project.authInfo into auth table.
+	return errors.EnsureStack(err)
+}
+
+// GetProject retrieves an entry from the core.projects table by project name.
+func GetProject(ctx context.Context, tx *pachsql.Tx, projectName string) (*pfs.ProjectInfo, error) {
+	return getProject(ctx, tx, "where name = $1", projectName)
+}
+
+// GetProjectByID is like GetProject, but retrieves an entry using the row id.
+func GetProjectByID(ctx context.Context, tx *pachsql.Tx, id int64) (*pfs.ProjectInfo, error) {
+	return getProject(ctx, tx, "where id = $1", id)
+}
+
+func getProject(ctx context.Context, tx *pachsql.Tx, where string, whereVal interface{}) (*pfs.ProjectInfo, error) {
+	row := tx.QueryRowContext(ctx, fmt.Sprintf("SELECT name, description, created_at FROM core.projects %s", where), whereVal)
+	project := &pfs.ProjectInfo{Project: &pfs.Project{}}
+	var createdAt time.Time
+	var err error
+	if err = row.Scan(&project.Project.Name, &project.Description, &createdAt); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	project.CreatedAt, err = types.TimestampProto(createdAt)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return project, nil
+}
+
+// UpdateProject updates all fields of an existing project entry in the core.projects table by name. If 'upsert' is set to true, UpdateProject()
+// will attempt to call CreateProject() if the entry does not exist. If updateTime is not provided, the current time will be used.
+func UpdateProject(ctx context.Context, tx *pachsql.Tx, project *pfs.ProjectInfo, upsert bool, updateTime ...*types.Timestamp) error {
```

Comment:
Hmm the `updateTime ...*types.Timestamp` is annoying. I'm guess you are using variadic arguments to support optional arguments. Ideally, `pfs.ProjectInfo` already has the `updated_at` field.